lib/rack/query_parser.rb

```diff
@@ -215,20 +211,37 @@ def params_hash_has_key?(hash, key)
       true
     end
 
-    def check_query_string(qs, sep)
-      if qs
-        if qs.bytesize > @bytesize_limit
-          raise QueryLimitError, "total query size (#{qs.bytesize}) exceeds limit (#{@bytesize_limit})"
-        end
+    def each_query_pair(qs, separator, unescaper = nil)
+      return enum_for(:each_query_pair, qs, separator, unescaper) unless block_given?
 
-        if (param_count = qs.count(sep.is_a?(String) ? sep : '&')) >= @params_limit
-          raise QueryLimitError, "total number of query parameters (#{param_count+1}) exceeds limit (#{@params_limit})"
-        end
+      return if !qs || qs.empty?
 
-        qs
+      if qs.bytesize > @bytesize_limit
+        raise QueryLimitError, "total query size (#{qs.bytesize}) exceeds limit (#{@bytesize_limit})"
+      end
+
+      pairs = qs.split(separator ? (COMMON_SEP[separator] || /[#{separator}] */n) : DEFAULT_SEP, @params_limit + 1)
+
+      if pairs.size > @params_limit
+        param_count = pairs.size + pairs.last.count(separator || "&")
+        raise QueryLimitError, "total number of query parameters (#{param_count}) exceeds limit (#{@params_limit})"
+      end
+
+      if unescaper
```

Comment:
I've gone for some small duplication here, so that the default-`unescaper` happy-path can avoid the whole `method(:unescape)` indirection

Reply:
I'm fine with this. What are your thoughts on deprecating the block argument to `parse_query` and the optional argument to this method, so we can eventually drop the duplicate code?